application/dash_application/views/bcl2barcode.py

```diff
@@ -28,47 +30,69 @@
     ]
 )
 
+def maybe_index2(s):
+    try:
+        return s.split("-")[1]
+    except IndexError:
+        return ""
+
 # There can be many unknown indices in each run. Only display top N
 MAX_UNKNOWN_DISPLAY = 30
 
-DATAVERSION = util.cache.versions(["bcl2fastq"])
-index = util.get_bcl2fastq_known()
-index_col = gsiqcetl.column.Bcl2FastqKnownColumn
+DATAVERSION = util.cache.versions(["bcl2barcode"])
+bcl2barcode = util.get_bcl2barcode()
+bcl2barcode_col = gsiqcetl.column.Bcl2BarcodeColumn
+pinery = util.get_pinery_samples()
+PINERY_COL = util.PINERY_COL
+
+barcode_expansions = pandas.DataFrame(columns=['Index', 'Sequence'])
+with open(os.getenv("BARCODES_STREXPAND"), 'r') as strexpand:
+    for l in strexpand.readlines():
+        x = l.split("\t")
+        for i in range(1,5):
+            barcode_expansions = barcode_expansions.append({'Index': x[0], 'Sequence': x[i].replace("\n", "")}, ignore_index=True)
+
+# Expand pinery to include 4 rows for every 1 10X barcode
+pinery_with_expanded_barcodes = pandas.merge(pinery, barcode_expansions, left_on='iusTag', right_on='Index', how='left')
+
+# Where there is no 10X barcode, fill in with the iusTag
+pinery_with_expanded_barcodes['Sequence'] = pinery_with_expanded_barcodes['Sequence'].fillna(pinery_with_expanded_barcodes['iusTag'])
+
+# Merge the expanded pinery with bcl2barcode to expand the bcl2barcode data the same way
+bcl2barcode_with_pinery = pandas.merge(bcl2barcode, pinery_with_expanded_barcodes, left_on=['Run Alias', 'Lane Number', 'Barcodes'], right_on=['sequencerRunName', 'laneNumber', 'Sequence'], how='left')
+
+# Failures to merge with the pinery data populate the 'Unknown' table
+unknown_data_table = bcl2barcode_with_pinery.loc[bcl2barcode_with_pinery['studyTitle'].isnull()]
+
+# Rows which merged successfully are the 'Known' table
+known_data_table = bcl2barcode_with_pinery.loc[bcl2barcode_with_pinery['studyTitle'].notnull()]
```

Comment:
When I look at Run 200504_A00469_0101_AH7MC7DSXY the 10X runs (GNLR) are not shown. They seem to go into the Unknown table.

Reply:
I took csvs of both data sources (which I can provide) and compared data for this run on Lane 4 (where all the GNLR info is). Everything in bcl2barcode for this run on that lane has 16-char barcodes. Since the expanded barcodes that go into the pinery df are 8-char, they fail to merge when creating bcl2barcode_with_pinery. 
Do the barcode expansions look like they're being established properly? Is merging on Barcodes and Sequence logical?

Reply:
Right. This is a general problem where the barcodes in Pinery OR in the run, could be shorter. I was wrong in recommending `startswith`. The comparison should be:
```
pinery[0:min(len(pinery), len(bcl2barcode)] == bcl2barcode[0:min(len(pinery), len(bcl2barcode)]
```

I'm not sure how that translates to pandas...

Reply:
Pushed attempts at implementing this, doesn't seem to work, Savo and I will talk about it this afternoon